cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/ArtifactHttpHandlerInternal.java

```diff
@@ -107,14 +111,24 @@ public void getArtifactBytes(HttpRequest request, HttpResponder responder,
                                @PathParam("namespace-id") String namespaceId,
                                @PathParam("artifact-name") String artifactName,
                                @PathParam("artifact-version") String artifactVersion,
-                               @QueryParam("scope") @DefaultValue("user") String scope) throws Exception {
+                               @QueryParam("scope") @DefaultValue("user") String scope,
+                               @DefaultValue("0") @HeaderParam(LAST_MODIFIED_HEADER) final String lastModified) throws Exception {
 
     NamespaceId namespace = validateAndGetScopedNamespace(Ids.namespace(namespaceId), scope);
     ArtifactId artifactId = new ArtifactId(namespace.getNamespace(), artifactName, artifactVersion);
     ArtifactDetail artifactDetail = artifactRepository.getArtifact(Id.Artifact.fromEntityId(artifactId));
     Location location = artifactDetail.getDescriptor().getLocation();
-    responder.sendContent(HttpResponseStatus.OK, new LocationBodyProducer(location),
-                          new DefaultHttpHeaders().add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_OCTET_STREAM));
+
+    HttpHeaders headers = new DefaultHttpHeaders()
+      .add(HttpHeaderNames.CONTENT_TYPE, HttpHeaderValues.APPLICATION_OCTET_STREAM)
+      .add(HttpHeaderNames.LAST_MODIFIED, location.lastModified());
+
+    if (Long.parseLong(lastModified) == location.lastModified()) {
```

Comment:
The `lastModified` is not a long. Please see https://datatracker.ietf.org/doc/html/rfc7232#section-3.3 and https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Modified-Since

Reply:
refactored it to use Date

Reply:
Make sure to use classes from the new java.time.* package instead of the old java.util.Date